File: mcp-code-intelligence-python/src/mcp_code_intel/analyzers/security/dataflow/data_flow_analyzer.py
"""
KSA-164: Data Flow Analyzer — Computes def-use chains from CFG + reaching definitions.
"""

from __future__ import annotations
from ..cfg.control_flow_graph import ControlFlowGraph
from .reaching_definitions import ReachingDefinitions
from ..types import DataFlowResult, DefUseChain, Definition


class DataFlowAnalyzer:
    def __init__(self) -> None:
        self._reaching_defs = ReachingDefinitions()

    def analyze(self, cfg: ControlFlowGraph) -> DataFlowResult:
        """Analyze data flow for a control flow graph."""
        reaching = self._reaching_defs.compute(cfg)
        all_defs = self._collect_all_definitions(cfg)
        chains = self._build_def_use_chains(cfg, reaching, all_defs)

        return DataFlowResult(
            reaching_defs=reaching,
            def_use_chains=chains,
            definitions=all_defs,
        )

    def _collect_all_definitions(self, cfg: ControlFlowGraph) -> list[Definition]:
        defs: list[Definition] = []
        def_id = 0
        for block in cfg.blocks:
            for var_def in block.get_definitions():
                defs.append(Definition(
                    variable=var_def.name,
                    line=var_def.line,
                    block_id=block.id,
                    id=def_id,
                ))
                def_id += 1
        return defs

    def _build_def_use_chains(
        self,
        cfg: ControlFlowGraph,
        reaching: dict[int, set[Definition]],
        all_defs: list[Definition],
    ) -> list[DefUseChain]:
        chains: list[DefUseChain] = []

        for defn in all_defs:
            uses: list[dict] = []

            for block in cfg.blocks:
                reaching_here = reaching.get(block.id, set())

                # Check if this definition reaches this block
                reaches = any(
                    d.variable == defn.variable and d.line == defn.line and d.block_id == defn.block_id
                    for d in reaching_here
                )
                if not reaches:
                    continue

                # Find uses of this variable in the block
                for use in block.get_uses():
                    if use.name == defn.variable:
                        uses.append({"line": use.line, "block_id": block.id})

            if uses:
                chains.append(DefUseChain(definition=defn, uses=uses))

        return chains

File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/security/dataflow/DataFlowAnalyzer.kt
/**
 * KSA-164: Data Flow Analyzer — Computes def-use chains from CFG + reaching definitions.
 */
package com.codeintel.analyzers.security.dataflow

import com.codeintel.analyzers.security.DataFlowResult
import com.codeintel.analyzers.security.DefUseChain
import com.codeintel.analyzers.security.Definition
import com.codeintel.analyzers.security.cfg.ControlFlowGraph

class DataFlowAnalyzer {
    private val reachingDefs = ReachingDefinitions()

    fun analyze(cfg: ControlFlowGraph): DataFlowResult {
        val reaching = reachingDefs.compute(cfg)
        val allDefs = collectAllDefinitions(cfg)
        val chains = buildDefUseChains(cfg, reaching, allDefs)
        return DataFlowResult(reachingDefs = reaching, defUseChains = chains, definitions = allDefs)
    }

    private fun collectAllDefinitions(cfg: ControlFlowGraph): List<Definition> {
        val defs = mutableListOf<Definition>()
        var id = 0
        for (block in cfg.blocks) {
            for (varDef in block.getDefinitions()) {
                defs.add(Definition(variable = varDef.name, line = varDef.line, blockId = block.id, id = id++))
            }
        }
        return defs
    }

    private fun buildDefUseChains(
        cfg: ControlFlowGraph,
        reaching: Map<Int, Set<Definition>>,
        allDefs: List<Definition>
    ): List<DefUseChain> {
        val chains = mutableListOf<DefUseChain>()

        for (def in allDefs) {
            val uses = mutableListOf<Map<String, Int>>()

            for (block in cfg.blocks) {
                val reachingHere = reaching[block.id] ?: continue
                val reaches = reachingHere.any {
                    it.variable == def.variable && it.line == def.line && it.blockId == def.blockId
                }
                if (!reaches) continue

                for (use in block.getUses()) {
                    if (use.name == def.variable) {
                        uses.add(mapOf("line" to use.line, "blockId" to block.id))
                    }
                }
            }

            if (uses.isNotEmpty()) {
                chains.add(DefUseChain(definition = def, uses = uses))
            }
        }

        return chains
    }
}
